[Shaune Ang] iP

shauneang opened the pull request:
### Duke Pro
> "Your mind is for having ideas, not holding them." - David Allen([source](https://dansilvestre.com/productivity-quotes))
DukePro frees your mind of having to remember things you need to do. It's, 

- text-based
- easy to learn
- ~~FAST~~*SUPER*FAST to use

All you need to do is, 

1. download it from here.
2. double-click it..
3. add your tasks.
4. let it manage your tasks for you :+1:

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java ruby
public class Main {
        public static void main(String[] args) {
                Application.launch(MainApp.class, args);
        }
}
```

Commit:
Add Gradle support

shauneang committed:
Duke echos commands entered by the user and exits when the user types bye

shauneang committed:
Duke echos commands entered by the user and exits when the user types bye

shauneang committed:
Level 2: Duke takes in list command and outputs all prior tasks given

shauneang committed:
Added try catch statement for input line

shauneang committed:
Abstracted task methods and added javadocs

shauneang committed:
Abstract out the ToDoList and Task from Duke

shauneang committed:
Added subtypes of Task (Todos, Deadline, Event)

shauneang committed:
added simple text Ui Testing

shauneang committed:
Added delete command

shauneang committed:
Error Handling for invalid commands

shauneang committed:
Create Delete command

shauneang committed:
Fixed JavaDocs Mistakes

shauneang committed:
Added Save Function for to do list tasks

shauneang committed:
testing

shauneang committed:
test

shauneang committed:
Level-8: Dates and Times

shauneang committed:
Level-8: fixed javadocs

shauneang committed:
Merge branch 'branch-Level-8'

shauneang committed:
Merging: fixing merge problems

shauneang committed:
add-gradle-support

shauneang committed:
add config folder and build.gradle file

shauneang committed:
A-Gradle: added checkstyle

shauneang committed:
added dummy test

shauneang committed:
A-MoreOOP: Abstracted out command, fileloadersaver, parser and ui

shauneang committed:
A-MoreOOP: Added javadocs

shauneang committed:
created duke package and duke.task package

shauneang committed:
corrected file path bug

shauneang committed:
A-JUnit: Created FileLoaderSaver Test and Command Test

shauneang committed:
A-Jar: Created a jar file for duke

shauneang committed:
A-CodingStandard: Checked CS2103T coding standards

shauneang committed:
A-Find: Added feature to search for tasks based on search string

shauneang committed:
changed test names to proper convention

shauneang committed:
Level-10: added JavaFX GUI

RezwanArefin01 commented on src/main/java/duke/FileLoaderSaver.java:
```diff
@@ -0,0 +1,49 @@
+    public List<String> loadFile() throws IOException{
+        //Find existing toDoList
+        if (!Files.exists(filePath))
+        {
```
'{' should be in the previous line.

RezwanArefin01 commented on src/main/java/duke/task/ToDos.java:
```diff
@@ -0,0 +1,33 @@
+ * duke.task.ToDos class creates todos
+ * @author Shaune Ang
+ */
+public class ToDos extends Task{
```
Should the class be named "ToDo" to match the other Task sub classes (in singular). 
Also there should be a space before '{'.

RezwanArefin01 commented on src/main/java/duke/task/ToDos.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    /**
+     * duke.task.ToDos constructor for task loaded from saved file
```
The JavaDoc comment should start with a verb. For example: `Create a new {@code ToDo} Task ...`

RezwanArefin01 commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,183 @@
+ * @author Shaune Ang
+ */
+public class Command {
+    static final String EXITWORD = "bye";
```
Is the default access modifier the correct choice?

RezwanArefin01 commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,183 @@
+                storage.writeToFile(toDoList.createTxtFile());
+                toDoList.displayListSize();
+                break;
+            case MARK:
```
Add a comment such as `// Fallthrough` to make it more explicit.

RezwanArefin01 commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,183 @@
+    private void changeMark(String command) {
+        String[] splitComm = command.split(" ");
+        String action = splitComm[0];
+        int index = Integer.parseInt(splitComm[1]) - 1;
```
What if `parseInt` throw an exception? 

Even though the number is checked with regex during parsing, I feel it might be better to handle the exception.
Or even better approach is to make the Parser also parse the index, and call this method with the index. This way all the parsing will be done in one place and the exceptions related to parsing can also be thrown from one place.

RezwanArefin01 commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,183 @@
+     */
+    private void changeMark(String command) {
+        String[] splitComm = command.split(" ");
+        String action = splitComm[0];
```
It might be better to `.strip()` the command and the split to skip extra white spaces. Upto you for the behavior you expect.

RezwanArefin01 commented on src/main/java/duke/ToDoList.java:
```diff
@@ -0,0 +1,165 @@
+ * @author Shaune Ang
+ */
+public class ToDoList {
+    ArrayList<Task> toDoList;
```
Is default access modifier the correct option? since you are exposing the functionality to add tasks through some provided methods, I think this can be `private`.

yuxuanleong replied:
Alternatively, a public `getTask` method can be added so that you can modify the tasks in other files.

RezwanArefin01 commented on src/main/java/duke/ToDoList.java:
```diff
@@ -0,0 +1,165 @@
+     * @throws IndexOutOfBoundsException
+     */
+    public void complete(int index) throws IndexOutOfBoundsException {
+        if (index >= toDoList.size()) {
```
What if `index` is negative? Same bellow in couple of places.

RezwanArefin01 commented on src/main/java/duke/ToDoList.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    
+    /** Returns todolist formatted for saving in duke.Duke.txt file in hard disk
```
There should be a newline after `**`.

RezwanArefin01 commented on src/main/java/duke/ToDoList.java:
```diff
@@ -0,0 +1,165 @@
+    public String createTxtFile() {
+        String result = "";
+        for (Task t : toDoList) {
+            if (t instanceof ToDos) {
```
Missed opportunity to use polymorphism here. You can add an abstract `toSaveString()` in `Task` which the sub-classes implement. Then it will be possible to just do `result += t.toSaveString() + "\n";` here. :smiley:

RezwanArefin01 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,50 @@
+    private Scanner myScanner;
+
+    /**
+     * duke.Ui constructor initiates scanner object
```
JavaDoc header should start with a verb. For example `Constructs a new {@code Ui} ...`.
Also there should be a `.` after the header comment. It is missing from a lot of other places in the code.

RezwanArefin01 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,50 @@
+
+    /**
+     * Reads and returns full command given by user
+     * @return full command given by user
```
`@param`, `@return` should be separated from the description by a newline.

RezwanArefin01 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+     */
+    public Deadline(String name, String deadline) {
+        super(name);
+        this.deadline = LocalDate.parse(deadline);
```
Should this be done in Parser class instead?

RezwanArefin01 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+     */
+    public void mark() {
+        String message;
+        if (status == true) {
```
`== true` in unnecessary.

RezwanArefin01 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,84 @@
+ */
+abstract public class Task {
+    private String name;
+    private Boolean status = false;
```
Maybe rename the variable to `isDone` or something similar.

RezwanArefin01 reviewed (commented):
Looks good. Some changes regarding coding standard and code quality suggested. :+1:

yuxuanleong commented on src/main/java/duke/Command.java:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    /**
+     * Constructor for command objects
+     * @param commandType type of command this command with execute
+     * @param fullCommand full user input command in String
```
Following the Javadoc comments convention would be better (empty line between description an parameter section)

yuxuanleong commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,45 @@
+    public static Command parse(String command) throws Exception {
+        if (command.equals("list")) {
+            return new Command(Command.CommandTypes.LIST, command);
+        }
+        else if (command.startsWith("todo")) { // create todos
```
`else if` should be right after the closing `}` of the previous `if`

yuxuanleong reviewed (commented):
Overall a decent job done 😄 ! Just some javadocs are slightly off and some improvements can be made regarding the OOP principles.

shauneang committed:
Change Duke GUI

Duke now uses JavaFX to make its GUI instead of CLI

Create JAR file so users can access Duke directly

Made code more OOP by abstracting the parser, TaskList, Command, FileLoaderSaver

shauneang committed:
A-Assertions: Added assertions in TaskList and Main Window

shauneang committed:
Merge pull request #1 from shauneang/branch-A-Assertions

Branch a assertions

shauneang committed:
A-CodeQuality: Improved on code quality

shauneang committed:
A-CodeQuality: Improved code quality of Command.java

shauneang committed:
Merge pull request #2 from shauneang/branch-A-CodeQuality

Branch a code quality

shauneang committed:
branch-B-Priority: added priority field into task

shauneang committed:
Merge pull request #3 from shauneang/branch-B-Priority

branch-B-Priority: added priority field into task

shauneang committed:
Edited formatting of code to fit with coding standard

shauneang committed:
Edited code to fit with coding standard

shauneang committed:
Abstract Dialog Box out and improved Ui of duke chatbot

shauneang committed:
A-UserGuide: Added User Guide and improved Ui

shauneang committed:
debug load file

shauneang committed:
add UI.png

shauneang committed:
added ReadMe.md

shauneang committed:
edit readme.md

shauneang committed:
edit readme.md

shauneang committed:
Added Ui image

shauneang committed:
test remove UI.png

shauneang committed:
adding back Ui.png

shauneang committed:
fixed duke.txt filepath
